episodes/instance-segmentation-and-measurements.md

```diff
@@ -0,0 +1,596 @@
+the radius of the footprint to control the amount of erosion. Try eroding
+the `semantic_seg` layer with different integer values for the radius.
+What radius do you need to ensure all nuclei are separate?
+
+```python
+from skimage.morphology import binary_erosion, ball
+
+# With radius = 1
+radius = 1
+eroded_mask = binary_erosion(semantic_seg, footprint = ball(radius))
+viewer.add_labels(eroded_mask, name = f'eroded ball {radius}')
+```
+
+:::::::::::::::::::::::::solution
+To test different values of radius, you can assign a different value to
+radius, e.g. `radius = 5` and rerun the second two lines from above. Or
+you can try with a Python [for loop](
+https://swcarpentry.github.io/python-novice-inflammation/05-loop.html)
+which enables us to test multiple values of radius quickly.
+```python
+# The for loop will repeat the indented lines of codes for each value
+# of radius in the list (1, 5, 10).
+for radius in 1, 5, 10:
+  eroded_mask = binary_erosion(semantic_seg, footprint = ball(radius = 1))
+  viewer.add_labels(eroded_mask, name = f'eroded ball {radius}')
+
+```
+
+![](fig/binary_mask_no_erosion.png){
+alt="A screenshot of a semantic segmentation mask before erosion."}
+The first image shows the mask without any erosion for comparison.
+
+![](fig/binary_mask_erosion_1.png){
+alt="A screenshot of a semantic segmentation mask eroded with a ball of
+radius 1."}
+Erosion with a radius of 1 makes a small difference, but the nuclei remain
+joined.
+
+![](fig/binary_mask_erosion_5.png){
+alt="A screenshot of a semantic segmentation mask eroded with a ball of
+radius 5."}
+Erosion with a radius of 5 makes a more noticeable difference,
+but some nuclei remain joined.
+
+![](fig/binary_mask_erosion_10.png){
+alt="A screenshot of a semantic segmentation mask eroded with a ball of
+radius 10."}
+Erosion with a radius of 10 separates all nuclei.
+
+
+:::::::::::::::::::::::::
+:::::::::::::::::::::::::
+
+Now we have separate nuclei, lets try creating instance labels
+again.
+
+```python
+# Remove the incorrect instance segmentation from the viewer
+viewer.layers.remove('instance_seg')
+
+eroded_semantic_seg = viewer.layers['eroded ball 10'].data
+# Create a new instace segmentation using the eroded mask
+instance_seg = label(eroded_semantic_seg)
+
+viewer.add_labels(instance_seg)
+
+print(f"Number of nuclei after erosion  = {instance_seg.max()}")
+```
+
+```output
+Number of nuclei after erosion  = 19
+```
+
+![](fig/instance_segmentation_eroded.png){
+alt="Instance segmentation on the eroded segmentation mask"}
+Looking at the image above, there are no longer any incorrectly joined
+nuclei. The absolute number of nuclei found hasn't changed much as the
+erosion process has removed some partial nuclei around the edges of the
+image.
+
+Performing any size or shape analysis on these nuclei will be flawed, as
+they are heavily eroded. We can largely undo much of the erosion by using
+the scikit-images [expand labels](
```

Comment:
This appears to be different from a dilation, correct?

Reply:
It's different to a dilation in that it won't create overlapping labels. My understanding is that it grows the bright (non zero) regions in the same way as a dilation but stops growing when it hits a neighbouring label. Maybe I can add something to the text to clarify this.

Reply:
Added note to text with 3aa7c77